### ISSUE TITLE
Problems with background sample processing, mosaic/mixup performance, single/multi-GPU accuracy gap and lack of multi-class keypoint detection

### ISSUE THREAD

**Issue by kaiwen0901** [Libre-YOLO/libreyolo#484]
When using the code from this repository for our project, we encountered several issues listed below and would greatly appreciate your assistance with relevant explanations and optimizations:
1. When training YOLO9 on a dataset with background images accounting for more than 80% of all samples, its mAP@50 is roughly 12% lower than YOLOv8. After removing all background images from the dataset and retraining, its accuracy becomes comparable to YOLOv8. Similar situations have been found in multiple datasets. We want to understand how background images are handled in the current code, and wonder if the relevant processing logic can be optimized to maintain stable training accuracy on datasets with an overabundance of background samples.
2. When training YOLO9 on our custom dataset, the model accuracy stays extremely low with barely any improvement during mosaic and mixup augmentation. Accuracy increases drastically only after we turn off mosaic and mixup. Could you confirm whether this phenomenon is expected behavior?
The picture below is a screenshot of 100 epochs of training：
<img width="897" height="524" alt="Image" src="https://github.com/user-attachments/assets/650927b2-1d6e-4d73-a6db-5aa2e38f82fc" />

<img width="782" height="510" alt="Image" src="https://github.com/user-attachments/assets/ce0f8866-1cfa-43b1-a63c-1385fe9af26c" />

<img width="721" height="506" alt="Image" src="https://github.com/user-attachments/assets/d6b227ec-cf01-458f-baae-4b78978c0326" />

<img width="711" height="507" alt="Image" src="https://github.com/user-attachments/assets/fbe6a2a4-998c-4248-9227-280552e01167" />

<img width="694" height="522" alt="Image" src="https://github.com/user-attachments/assets/77766182-09b1-4d33-b343-21ca9c16d0c3" />

3. Under the exact same dataset and identical training configurations, we observed inconsistent final accuracy between single-GPU and multi-GPU training.

<img width="1375" height="557" alt="Image" src="https://github.com/user-attachments/assets/f5e15735-be6e-4ca9-a53d-21a6c0ef5902" />

4. All keypoint detection models implemented in the repo only support single-class detection. Would it be feasible to add code support for multi-class keypoint detection?

**Comment by EHxuban11** [comment]
Hello @kaiwen0901 , thank you for the time you took to report this bugs, the experiments you did facilitated my debugging a lot. 

Regarding mosaic/mixup, the bugs are confirmed. Mixup was dropping 100% of the second image's labels (making them backgrounds). Mosaic was mis-scaling objects. Both are fixed now on the dev branch. 

Rregarding the background images problem, I could also reproduce your problem training some models. Your background problem was also magnified hy the Mixup bug which was adding even more backgrounds, but even with fixed Mixup the problem still exists. There is no bug in how backgrounds are handled, empty label images are kept as negatives exactly like the reference implementation (https://github.com/MultimediaTechLab/YOLO). In my opinion, the CV engineer is the one in charge of finding the right ratio of background images to labelled images, LibreYOLO already ships a LibreDoctor diagnostics toolkit that would have flagged the high backgrouund ratio. From LibreYOLO's side, a warning could be issued during the training using the diagnostics toolkit to say if the background image percentage is too high, but it won't subsample them automatically. 

Regarding the GPU gap, thank you for reporting this as well. Currently i don't have two GPU's so I couldn't reproduce it. But there was a bug in the code regarding the trainer multiplying the loss by the number of GPU's which makes the learning rate bigger than what it should have. In your case LR was 4 times higher. This problem is also fixed, but I would appreciate if you could run the experiment again.

Regarding the Multi-class keypoints, all the models are currently single-class as you could see. That's also true of their upstream sources (upstream yolonas, upstream rf-detr, etc.). It is possible to modify the models experimentally by changing the head, loss, loader and checkpoint schema, but each model's difficulty differs. I can add multi-class keypoint to the roadmap but it's gonna take some time. What model would you like to be able to use multi class keypoints on?

Thank you again for your detailed reports, they really made the library better.

**Comment by EHxuban11** [comment]
Hello @kaiwen0901 , I've been investigating multi class keypoint detectors and there are two types: same skeleton for different classes (for example multiple different animals but they have the same set of keypoints) and different skeletons.

The second seem extremly hard and esentially an unsolved problem although I might be wrong. Could you clarify what is your need?  Have you succesfully used model for this use case in the past and point me to them? Thank you.

**Comment by kaiwen0901** [comment]
Hello, we have conducted tests using the latest code on the dev branch, and here are our test results and requirements for multi-class keypoints:
1.We retrained on the dataset where we previously encountered performance degradation, comparing training results before and after removing background images. The accuracy gap has been greatly narrowed: the difference of mAP50 is only about 0.5%, and mAP50-95 differs by roughly 2%.
2.We ran multiple repeated training trials. The accuracy curve during mosaic augmentation now behaves completely normally without stagnation.
Partial logs from one training process are attached below.

<img width="520" height="450" alt="Image" src="https://github.com/user-attachments/assets/6fc23845-056d-4e86-af55-611320544486" />

<img width="572" height="505" alt="Image" src="https://github.com/user-attachments/assets/cf2af260-9b69-47dc-82b7-a1c91e41ce93" />

<img width="582" height="509" alt="Image" src="https://github.com/user-attachments/assets/dc3326d2-7cf9-43a1-a3be-2c45082ddbb6" />

<img width="688" height="565" alt="Image" src="https://github.com/user-attachments/assets/8dea2252-f48a-4a12-9af7-cda28aa16654" />

<img width="664" height="565" alt="Image" src="https://github.com/user-attachments/assets/185150d2-aae2-4479-bf1f-672d254e00dd" />

<img width="661" height="563" alt="Image" src="https://github.com/user-attachments/assets/fcc0fe3c-2a88-47e5-ad6c-2d3255c8d5f6" />

3.The accuracy between single-GPU and multi-GPU training is nearly consistent after the fixes.

<img width="1525" height="855" alt="Image" src="https://github.com/user-attachments/assets/9b91e8d7-6366-4cb8-86fb-26ba78e2d36e" />

4.For our project, we mainly test the yolonas-pose model.
Regarding multi-class keypoint detection: our keypoint task is not human pose estimation, but meter detection. We need to predict the central keypoint of each meter while classifying different meter types. All meter categories share exactly one central keypoint, which falls into your first scenario (all classes have identical skeleton/keypoint layout).
In addition, we have verified that Ultralytics YOLO (https://github.com/ultralytics/ultralytics) natively supports multi-class keypoint detection. I will attach our sample training YAML configuration file for your reference below.

<img width="621" height="586" alt="Image" src="https://github.com/user-attachments/assets/1bf55c44-869d-4490-867b-d6ed612ef0ee" />

<img width="595" height="508" alt="Image" src="https://github.com/user-attachments/assets/95f04866-270a-4aad-b22e-15b6181e871d" />

<img width="451" height="577" alt="Image" src="https://github.com/user-attachments/assets/f73d27b5-624d-4f63-8daa-a270652f4b17" />

<img width="405" height="356" alt="Image" src="https://github.com/user-attachments/assets/e4f4b92f-d569-4069-831f-ac0513be9316" />

**Comment by EHxuban11** [comment]
Hello @kaiwen0901 , 

Thank you for taking the time to run the experiments and reporting. 

Regarding the background images, glad to hear that the gap narrowed to 0.5. Regarding mosaic and mixup thanks for confirming they are fixed.

For single vs multi GPU gap I added one more improvement (SyncBatchNorm) after your test which could completely close the gap. 

I took the time to add multi-class keypoints for yolonas-pose and it is merged on the dev branch (#530 ). It covers exactly your case where there are multiple classes that share the same keypoint skeleton. The config like the one you showed should work directly.  If you can test it on your meter dataset I would love to hear how it goes.

**Comment by kaiwen0901** [comment]
Thank you so much for addressing all these issues promptly. We have finished testing the new multi-class keypoint feature on our meter dataset, and the training results work perfectly. We will provide timely feedback if we encounter any new problems in subsequent use. We really appreciate your continuous attention and support to this issue.

**Comment by EHxuban11** [comment]
Hello @kaiwen0901 , you are welcome, it has been very nice to work with you on improving these issues. Thank you for the detailed reports each time and your attentio to this matter. Have a nice week!

**Comment by kaiwen0901** [comment]
Hello @EHxuban11 , regarding YOLONAS-Pose, we found another long-standing issue:
When training with the same batch_size setting, the memory usage of each GPU in multi-GPU training is almost the same as that in single-GPU training. This suggests the total batch size is not split evenly across every GPU as expected during distributed training.
In addition, an error prompt like the one in the attached screenshot occasionally appears during training, but it does not impact the training process at all.

<img width="1164" height="574" alt="Image" src="https://github.com/user-attachments/assets/7d04cd7b-f988-4a58-9b0f-eeaf21696832" />

<img width="1220" height="773" alt="Image" src="https://github.com/user-attachments/assets/257b3a87-9550-4bbd-b71f-137e73fd62aa" />

**Comment by EHxuban11** [comment]
Hello @kaiwen0901 , thank you for reporting this. Both findings are real bugs. 

Regarding the batch split the problem is that yolo nas pose has it's own dataloader because the data in pose is different, and this data loader did not split correctly as you observed in the bug. The same bug existed in other families that have their own data loader or override wrong such as DEIM and DFINE. Now it is fixed for all of them (https://github.com/LibreYOLO/libreyolo/pull/605).

Regarding the errors in the terminal, the 4 GPU's were trying to read and write on the same predictions.json. Although it might have been harmeless in your case it could have been very problematic, so thank you for pointing to that as well. It is also fixed, now, only one process accesses the json file. 

Both fixes are merged in #605.  I also tested the fixes by renting 2 GPUs in vast.ai and it seems to work fine.

Please let me know if you find more issues. These are so valuable for improving the library. 

BR, Xuban

**Comment by kaiwen0901** [comment]
We have run tests to verify the two previously reported issues, and both problems are fully resolved now. Thank you very much for your hard work on the fixes.

**Comment by EHxuban11** [comment]
Hello @kaiwen0901 , super happy to hear that. As always if you find other problems and have the time to report I would appreciate that, together we have pushed the librarie's quality forward significantly.

Have a nice  weekend. 
BR,
Xuban
